src/ffi_helpers.c

```diff
@@ -0,0 +1,95 @@
+#include <stdbool.h>
+
+char _sh_ptr_read_c_char(char *ptr, int offset) {
+  return *(char *)(ptr + offset);
+}
+void _sh_ptr_write_c_char(char *ptr, int offset, char v) {
+  *(char *)(ptr + offset) = v;
+}
+signed char _sh_ptr_read_c_schar(char *ptr, int offset) {
+  return *(signed char *)(ptr + offset);
+}
+void _sh_ptr_write_c_schar(char *ptr, int offset, signed char v) {
+  *(signed char *)(ptr + offset) = v;
+}
+unsigned char _sh_ptr_read_c_uchar(char *ptr, int offset) {
+  return *(unsigned char *)(ptr + offset);
+}
+void _sh_ptr_write_c_uchar(char *ptr, int offset, unsigned char v) {
+  *(unsigned char *)(ptr + offset) = v;
+}
+short int _sh_ptr_read_c_short(char *ptr, int offset) {
+  return *(short int *)(ptr + offset);
+}
+void _sh_ptr_write_c_short(char *ptr, int offset, short int v) {
+  *(short int *)(ptr + offset) = v;
+}
+short unsigned int _sh_ptr_read_c_ushort(char *ptr, int offset) {
+  return *(short unsigned int *)(ptr + offset);
+}
+void _sh_ptr_write_c_ushort(char *ptr, int offset, short unsigned int v) {
+  *(short unsigned int *)(ptr + offset) = v;
+}
+int _sh_ptr_read_c_int(char *ptr, int offset) {
+  return *(int *)(ptr + offset);
```

Comment:
This is tricky. Type punning is risky to begin with, but the problem here is that there is no guarantee that the pointer is correctly aligned. That is why the static_h.h pointer functions have an API that ensures it. It wasn't just a whim 😀

This actually matters on Arm32 with 64-bit values. See this: https://godbolt.org/z/6fnTsGzTh

The function read4() literally contains incorrect code because the address must be aligned. read5() is what it is supposed to look like.

I am not sure what the best solution here is. Declare aligned and unaligned versions, because the offset is constant and the generator knows? Use the builtin_constant_p() trick in the function, check the alignment and do different things? Always use memcpy()?